src/machines/java/com/enderio/machines/client/gui/widget/EnergyTextboxWidget.java

```diff
@@ -0,0 +1,104 @@
+package com.enderio.machines.client.gui.widget;
+
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.components.EditBox;
+import net.minecraft.network.chat.Component;
+
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+public class EnergyTextboxWidget extends EditBox {
+    private final Supplier<Integer> maxEnergy;
+    
+    public EnergyTextboxWidget(Supplier<Integer> maxEnergy, Font font, int x, int y, int width, int height, Component message) {
+        super(font, x, y, width, height, message);
+        this.maxEnergy = maxEnergy;
+        setCanLoseFocus(true);
+    }
+	
+	@Override
+    public void setCanLoseFocus(boolean canLoseFocus) {
+        super.setCanLoseFocus(canLoseFocus);
+    }
+
+    //Input only accepts numerals
+    @Override
+    public boolean charTyped(char codePoint, int modifiers) {
+
+        if (codePoint >= 48 && codePoint <= 57) {
+            int cursorPos = getCursorPosition() + 1;
+            super.charTyped(codePoint, modifiers);
+            setValue(formatEnergy(getValue()));
+            moveCursorTo(cursorPos);
+        }
+
+        return false;
```

Comment:
yes.
Because I want only digits to be entered (0,1, ..., 9)
Any other character cannot be entered.
It could be improved if we think we need to do so by allowing prefixes like k = 1000 etc...